File: bot/indicators.py
"""Chỉ báo kỹ thuật. Công thức khớp TradingView để số liệu đối chiếu được."""
from __future__ import annotations

import numpy as np
import pandas as pd


def rma(x: pd.Series, period: int) -> pd.Series:
    """Wilder's smoothing (hàm ta.rma của TradingView).

    Điểm quan trọng: giá trị đầu tiên được SEED bằng SMA của `period` phần tử
    đầu, sau đó mới làm mượt đệ quy. Nếu dùng thẳng
    `ewm(alpha=1/period, adjust=False)` thì kết quả lệch tới hơn 10 điểm RSI
    so với TradingView vì nó seed bằng phần tử đầu tiên.
    """
    v = x.to_numpy(dtype=float)
    out = np.full(len(v), np.nan)

    # Bỏ qua các NaN ở đầu chuỗi (ví dụ phần tử đầu của diff)
    valid = np.flatnonzero(~np.isnan(v))
    if len(valid) < period:
        return pd.Series(out, index=x.index)
    s = valid[0]

    acc = v[s:s + period].mean()
    out[s + period - 1] = acc
    for i in range(s + period, len(v)):
        acc = (acc * (period - 1) + v[i]) / period
        out[i] = acc
    return pd.Series(out, index=x.index)


def rsi(close: pd.Series, period: int = 14) -> pd.Series:
    """RSI(period) theo Wilder — khớp TradingView."""
    delta = close.astype(float).diff()
    gain = delta.clip(lower=0.0)
    loss = (-delta).clip(lower=0.0)

    avg_gain = rma(gain, period)
    avg_loss = rma(loss, period)

    out = np.full(len(close), np.nan)
    ag, al = avg_gain.to_numpy(), avg_loss.to_numpy()
    mask = ~np.isnan(ag) & ~np.isnan(al)
    # avg_loss = 0 nghĩa là không có nến giảm nào -> RSI = 100
    zero = mask & (al == 0)
    norm = mask & (al != 0)
    out[zero] = 100.0
    out[norm] = 100.0 - 100.0 / (1.0 + ag[norm] / al[norm])
    return pd.Series(out, index=close.index)


def bollinger(close: pd.Series, period: int = 20, num_std: float = 2.0):
    """Trả (middle, upper, lower). ddof=0 giống TradingView."""
    mid = close.rolling(period).mean()
    sd = close.rolling(period).std(ddof=0)
    return mid, mid + num_std * sd, mid - num_std * sd


def sma(close: pd.Series, period: int = 50) -> pd.Series:
    return close.rolling(period).mean()


def crossed_above(series: pd.Series, level: float) -> bool:
    """Nến trước < level và nến hiện tại > level."""
    if len(series) < 2:
        return False
    prev, cur = series.iloc[-2], series.iloc[-1]
    return bool(pd.notna(prev) and pd.notna(cur) and prev < level < cur)


def crossed_below(series: pd.Series, level: float) -> bool:
    if len(series) < 2:
        return False
    prev, cur = series.iloc[-2], series.iloc[-1]
    return bool(pd.notna(prev) and pd.notna(cur) and prev > level > cur)

File: src/bot/indicators.ts
/**
 * Technical Indicators for VN100 Bot.
 * Formulas are calibrated to match TradingView exactly.
 */

/**
 * Wilder's smoothing (equivalent to ta.rma in TradingView).
 * Critical: The first value is seeded using the SMA of the first `period` elements,
 * and then smoothed recursively.
 */
export function rma(values: number[], period: number): (number | null)[] {
  const n = values.length;
  const out: (number | null)[] = new Array(n).fill(null);

  // Find first non-NaN/valid numbers
  let validStart = -1;
  for (let i = 0; i < n; i++) {
    if (!Number.isNaN(values[i]) && values[i] !== null && values[i] !== undefined) {
      validStart = i;
      break;
    }
  }

  if (validStart === -1 || n - validStart < period) {
    return out;
  }

  // Seed with SMA
  let sum = 0;
  for (let i = validStart; i < validStart + period; i++) {
    sum += values[i];
  }
  let acc = sum / period;
  out[validStart + period - 1] = acc;

  for (let i = validStart + period; i < n; i++) {
    acc = (acc * (period - 1) + values[i]) / period;
    out[i] = acc;
  }

  return out;
}

/**
 * RSI(period) using Wilder's RMA — matches TradingView.
 */
export function calculateRsi(close: number[], period: number = 14): (number | null)[] {
  const n = close.length;
  const out: (number | null)[] = new Array(n).fill(null);
  if (n < period + 1) return out;

  const gain: number[] = new Array(n).fill(0);
  const loss: number[] = new Array(n).fill(0);

  for (let i = 1; i < n; i++) {
    const diff = close[i] - close[i - 1];
    if (diff > 0) gain[i] = diff;
    else if (diff < 0) loss[i] = -diff;
  }

  const avgGain = rma(gain, period);
  const avgLoss = rma(loss, period);

  for (let i = 0; i < n; i++) {
    const ag = avgGain[i];
    const al = avgLoss[i];
    if (ag === null || al === null) {
      out[i] = null;
      continue;
    }
    if (al === 0) {
      out[i] = 100.0;
    } else {
      out[i] = 100.0 - 100.0 / (1.0 + ag / al);
    }
  }

  return out;
}

/**
 * Bollinger Bands calculation: middle, upper, lower. ddof = 0 (population stddev).
 */
export function calculateBollinger(
  close: number[],
  period: number = 20,
  numStd: number = 2.0
): { middle: (number | null)[]; upper: (number | null)[]; lower: (number | null)[] } {
  const n = close.length;
  const middle: (number | null)[] = new Array(n).fill(null);
  const upper: (number | null)[] = new Array(n).fill(null);
  const lower: (number | null)[] = new Array(n).fill(null);

  for (let i = period - 1; i < n; i++) {
    let sum = 0;
    for (let j = i - period + 1; j <= i; j++) {
      sum += close[j];
    }
    const mean = sum / period;

    let varianceSum = 0;
    for (let j = i - period + 1; j <= i; j++) {
      const diff = close[j] - mean;
      varianceSum += diff * diff;
    }
    const std = Math.sqrt(varianceSum / period); // ddof = 0

    middle[i] = mean;
    upper[i] = mean + numStd * std;
    lower[i] = mean - numStd * std;
  }

  return { middle, upper, lower };
}

/**
 * Simple Moving Average (SMA)
 */
export function calculateSma(close: number[], period: number = 50): (number | null)[] {
  const n = close.length;
  const out: (number | null)[] = new Array(n).fill(null);

  for (let i = period - 1; i < n; i++) {
    let sum = 0;
    for (let j = i - period + 1; j <= i; j++) {
      sum += close[j];
    }
    out[i] = sum / period;
  }

  return out;
}

export function isCrossedAbove(series: (number | null)[], level: number): boolean {
  if (series.length < 2) return false;
  const prev = series[series.length - 2];
  const cur = series[series.length - 1];
  return prev !== null && cur !== null && prev < level && cur >= level;
}

export function isCrossedBelow(series: (number | null)[], level: number): boolean {
  if (series.length < 2) return false;
  const prev = series[series.length - 2];
  const cur = series[series.length - 1];
  return prev !== null && cur !== null && prev > level && cur <= level;
}
